### ISSUE TITLE
Rule S2259: False positive on switch statements

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#2338]
### Description
The rule does not recognize which cases of a switch statement are safe (all except the one that contains the `null` case).

### Repro steps
``` Csharp
using System;

namespace SonarAnalyzer.Experiments.CSharp
{
    public class S2259
    {
        public ConsoleColor Color { get; set; }

        public void Method1(S2259 obj)
        {
            switch (obj?.Color)
            {
                case null:
                    Console.ForegroundColor = obj.Color; //not compliant
                    break;
                case ConsoleColor.Red:
                    Console.ForegroundColor = obj.Color; //compliant
                    break;
                default:
                    Console.WriteLine($"Color {obj.Color} is not supported."); //compliant
                    break;
            }
        }

        public void Method2(S2259 obj)
        {
            switch (obj?.Color)
            {
                case ConsoleColor.Red:
                    Console.ForegroundColor = obj.Color; //compliant
                    break;
                default:
                    Console.WriteLine($"Color {obj.Color} is not supported."); //not compliant
                    break;
            }
        }
    }
}
```

### Expected Result
Only the `null` cases should report S2259.

### Related information
* SonarC# Version 7.11.0.8083

**Comment by andrei-epure-sonarsource** [comment]
Another example [from the community](https://community.sonarsource.com/t/s2259-null-on-at-least-one-execution-path-despite-case-null/10192)

>This raises S2259 - "myObject is null on at least one execution path" in the default case (for the myObject.EnumValue dereferencing). But if myObject was null, then myObject?.EnumValue would also be null and the switch would hit the null case, throwing an exception. So default is unreachable if myObject is null.

```
void Foo()
{
    var myObject = GetObjectFromOutside();

    switch (myObject?.EnumValue)
    {
        case MyEnum.One:
            throw new MyException("Invalid value.");
        case MyEnum.Two:
        case null:
            throw new MyException("Invalid or null value.");
        case MyEnum.Three:
            break;
        default:
            throw new ArgumentOutOfRangeException(
                nameof(myObject.EnumValue),
                $"Unexpected value ${myObject.EnumValue}");
    }
}
```

**Comment by andrei-epure-sonarsource** [comment]
This seems highly related to #1183
